File: example/femasApiVisitor.cpp
#include "USTPFtdcUserApiStruct.h"
#include <pods_serdes_gen/messages.hpp>
#include <iostream>

struct StructVisitor
{
    void StartStruct(const std::string &structName)
    {
        std::cout << "Start of Struct: " << structName << std::endl;
        mStructName = structName;
    }

    void EndStruct() const
    {
        std::cout << "End of Struct" << std::endl;
    }

    template <typename T>
    void Member(const std::string &memberName, const T &member) const
    {
        std::cout << "Member: " << memberName << " in " << mStructName << " value is: " << member << std::endl;
    }

    std::string mStructName;
};

int main(int argc, char *argv[])
{
    CUstpFtdcRspUserLoginField data{};
    std::cout << "Before Change: " << pods_serdes_gen::to_string(data) << "\n" << std::endl;

    char strData[10];
    auto j = pods_serdes_gen::to_json(data);
    j["UserID"] = "123";
    j["FemasLifeCycle"] = 100;
    j["DataCenterID"] = 100;
    pods_serdes_gen::from_json(j, data);
    std::cout << "After Change: " << pods_serdes_gen::to_string(data) << "\n" << std::endl;

    std::cout << "Member visitor:" << std::endl;
    pods_serdes_gen::visit(StructVisitor{}, data);

    return 0;
}

File: example/femasApiVisitor.py
from api_femas import messages as FemasMessages
import json

def main():
    data = FemasMessages.CUstpFtdcRspUserLoginField()
    data.TradingDay = '20180101'
    data.UserID = '123'
    data.FemasLifeCycle = 100
    print("Before change: ", data)

    # Serialize
    j = json.dumps(vars(data))
    print("Serialized: ", j)

    # Deseiralize
    d = json.loads(j)
    d["UserID"] = "456"
    data = FemasMessages.CUstpFtdcRspUserLoginField(**d)
    print("After change: ", data)

if __name__ == "__main__":
    main()
